Add first template for a Helm resource

Gabriella439 opened the pull request:
This is an initial attempt to translate Helm idioms into Dhall idioms, using
an example resource from the Jenkins chart.

Gabriella439 committed:
Add first template for a Helm resource

This is an initial attempt to translate Helm idioms into Dhall idioms, using
an example resource from the Jenkins chart.

Gabriel439 commented:
There were a few design decisions that I want to explicitly call out here:

* The template only accepts the minimum record fields necessary rather than expecting the entire value

  I went with this approach to make templates easier to reuse and test in isolation

* I factored out the logic for templating labels

  This is cargo-culted between templates so I'm pretty confident this is the correct thing to do rather than to blindly follow the template as it was written

Other than that there is pretty much a one-to-one correspondence between the Helm template and the Dhall template

Gabriel439 commented:
Also, here is the original Helm template for comparison:

* [./stable/jenkins/templates/jcasc-config.yaml](https://github.com/helm/charts/blob/cedffd5231e833f72bceb80592527b03154ab21b/stable/jenkins/templates/jcasc-config.yaml)

Gabriel439 commented:
Actually, projecting out the fields of interest for nested records is very clumsy at the moment, so I plan to change this to instead take a uniform `Value` type for all templates related to a chart until it's easier to project out nested records.

Gabriella439 committed:
Use a uniform `Values` type for all templates

... just for better ergonomics in the top-level `chart.dhall` since it's
currently difficult to project out a subset of nested record fields

Gabriella439 committed:
More refactors to clean things up

* Add a new `utils.optional` utility
* Skip the use of `$` as an identifier
